NiL.JS/BaseLibrary/Console.cs

```diff
@@ -1,68 +1,395 @@
+            double elapsed = 0.0;
+            if (_timers.ContainsKey(label))
             {
-                var message = f.ToString();
-                message = message.Substring(message.IndexOf("=>") + 2).Trim();
-                System.Console.Error.WriteLine(message + " not equals " + sample);
+                _timers[label].Stop();
+                elapsed = (double)_timers[label].ElapsedTicks / Stopwatch.Frequency * 1000.0;
+                _timers.Remove(label);
             }
+            
+            if (label != "")
+                label += ": ";
+            LogMessage(LogLevel.Info, label + Tools.DoubleToString(System.Math.Round(elapsed, 10)) + "ms");
+
+            return JSValue.undefined;
         }
 
-        public static JSValue error(Arguments args)
+        
+
+        internal static List<console> _consoles;
+        internal static void AddConsole(console console)
         {
-            for (var i = 0; i < args.length; i++)
+            if (_consoles == null)
+                _consoles = new List<console>(4);
+            _consoles.Add(console);
+        }
+        internal static void RemoveConsole(console console)
+        {
+            if (_consoles == null)
+                _consoles = new List<console>(4);
+            _consoles.Add(console);
+        }
+        public static console GetConsole(GlobalContext globalContext)
+        {
+            if (_consoles == null)
+                return null;
+            for (int i = 0; i < _consoles.Count; i++)
             {
-                if (i > 0)
-                    System.Console.Error.Write(" ");
-                var r = args[i].ToString();
-                System.Console.Error.Write(r);
+                if (_consoles[i]._globalContext == globalContext)
+                    return _consoles[i];
             }
-            System.Console.Error.WriteLine();
             return null;
         }
+
+        public static JSValue CreateConsoleObject(GlobalContext globalContext)
+        {
+            return CreateConsoleObject(CreateConsole(globalContext));
+        }
+        public static JSValue CreateConsoleObject(GlobalContext globalContext, TextWriter logTw, TextWriter infoTw, TextWriter warnTw, TextWriter errorTw)
+        {
+            return CreateConsoleObject(CreateConsole(globalContext, logTw, infoTw, warnTw, errorTw));
+        }
+        public static JSValue CreateConsoleObject(console console)
+        {
+            return console.ConsoleObject;
+        }
+
+        public static console CreateConsole(GlobalContext globalContext)
+        {
+            console c = GetConsole(globalContext);
+            if (c != null)
+                return c;
+
+            c = new console(globalContext);
+            AddConsole(c);
+
+            return c;
+        }
+        public static console CreateConsole(GlobalContext globalContext, TextWriter logTw, TextWriter infoTw, TextWriter warnTw, TextWriter errorTw)
```

Comment:
Dnt wrt bbrvtd nms fr pblc mmbrs.
Users often do not understand what this means

Sometimes other programmers too.